app/services/aws/sts/assume_role_web_identity_service.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module Aws
+  module Sts
+    class AssumeRoleWebIdentityService
+      class << self
+        def call(params = {})
+          new(params).call
+        end
+      end
+
+      class TokenNotFoundError < StandardError
+      end
+
+      def initialize(params)
+        @params = params
+      end
+
+      def call
+        raise TokenNotFoundError, "web_identity_token_file was not found" unless web_identity_token_file_exists?
+
+        cached_call { Aws::AssumeRoleWebIdentityCredentials.new(params) }
+      end
+
+      private
+
+      attr_reader :params
+
+      def cached_call
+        cached_value = ::Rails.cache.read(cache_key)
+        return cached_value if cached_value.present?
+
+        sts_response = yield
+
+        ::Rails.cache.write(cache_key, sts_response, expires_in: cache_expires_in(sts_response.expiration))
```

Comment:
shouldn't we use just some class attribute/variable instead of using memcached? Do you see any benefits with that?

Reply:
I don't quite understand the suggestion of class attribute/variable 😬 
I added this cache so that we don't call AWS more than we need. Without it, each time we call AWS (for S3), we'll first have to call STS to get the temporary credentials. With the cache, we won't perform calls to STS until the cache expires, which's at the same time the temporary credentials expire.

Reply:
The credentials we get from `Aws::AssumeRoleWebIdentityCredentials` are temporary. They become invalid after 1 hour (by default) and, when that happens, we need to request for credentials again, using the token that is refreshed after X time as well.
Now, by adding cache, we'll only hit AWS again when we know the credentials will expire, as STS tells us the expiration date/time.
Those credentials will be used each time we have to communicate with S3 to fetch the assets we have stored there. If I understood your suggestion correct, by keeping the state of the credentials in a variable, we'd be able to use it only until the transaction finishes. Once a new transaction starts (page refresh?), we'd have to call AWS again.

Reply:
On https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/AssumeRoleWebIdentityCredentials.html I read:
> An auto-refreshing credential provider that assumes a role via STS::Client#assume_role_with_web_identity.

Are you sure it expires and new object has to be created?

Reply:
> Are you sure it expires and new object has to be created?

I'm sure it expires (check [here](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/STS/Client.html#assume_role_with_web_identity-instance_method) please) now that the new object has to be created is debatable. If we can keep the same object all the time, it'll be able to automatically go to AWS to fetch new credentials in case those expire.
Let me try to do something here... I'm thinking about a singleton 🤔

Reply:
yes, token expires. Just I was left with the impression that the credentials provider will silently handle that under the hood.

Would be nice to have it as a singleton of some sort. This is what I mean to keep it into a class instance variable.

Reply:
>Just I was left with the impression that the credentials provider will silently handle that under the hood

And it does indeed. And in quite a nice way... When we're using the credentials provided by `assume_role_with_web_identity`, if those are expired, the object makes sure to retrieve new credentials.
I added another commit (06d1aff) to remove the cache (yay!) in favour of a class that'll keep the `Aws::AssumeRoleWebIdentityCredentials` instance in memory. With this, unless that application instance changes, we should always be using the same object :v:

Thank you, @akostadinov! ... Could you please double check again 😬 ?